### PR TITLE
Misc fixes and improvements

### PR DESCRIPTION
**Pull request by skal65535**
### Fixes

- **EncodeYUV420**: the clipped path read the U plane with `v_step_` — wrong output whenever the U and V strides differ.
- **dichotomy**: `StoreRunLevels()` ignored `CheckBuffers()` and wrote past `all_run_levels_` once an allocation failed.
- **bit_writer**: `Reserve()` left `buf_` dangling into memory the sink had just released.
- **SjpegEncode**: `EncoderFactory()` returns null for an unknown `yuv_mode` or on allocation failure, and the result was dereferenced unchecked.
- **Encoder**: `param.memory` was installed after construction, so blocks allocated by sub-class constructors were released by the wrong manager.
- **EncodeNV12/NV21**: arguments were validated after the base class had already used `output`, and the `uv_stride` bound was too small by half.
- **Encoder/SOF**: dimensions above 65535 were silently truncated into the SOF marker — 70000x2 encoded "successfully" as 4464x2.
- **SjpegDimensions**: rejected the null width/height that the header documents as optional, and scanned chunks with pointers built past the end of the buffer.
- **SjpegRiskiness**: the gray fraction was divided by `width * height` although the loop visits `(width-1) * (height-1)` samples, and flat tinted colors fell into the gray bucket.
- **HeaderSize**: charged two quantization matrices where `WriteDQT()` writes only one.
- **yuv_convert**: two `static bool done` guards, torn when two threads encode at once.

### Quality

- **tests**: self-contained unit-test binary, no external dependency — `make unit_test`, or `ctest` with CMake; the linux and macos CI jobs were building the tests without running them, and now run them.
- **ci**: pin `actions/checkout` to a commit hash instead of a movable tag (same version).
- **misc**: clamp the compression method instead of asserting on it, add a missing include, fix two wrong comments.

Each fix comes with the unit test that catches it.

### Note for review

Two commits may change the bitstream at byte level: **HeaderSize** (only
`YUV_400 + TARGET_SIZE`) and **SjpegRiskiness** (`YUV_AUTO`'s choice on gray
and flat tinted pictures, which is the point of the fix).

Please rebase-merge rather than squash, so the fixes stay bisectable.

### CODE REVIEW

File: src/dichotomy.cc
Comment (hasinoff) [inline src/dichotomy.cc]: **⚠️ Bug / MSAN Use-of-Uninitialized-Value on Early Search Failure**

When running `TEST(AllocationFailure)` under MemorySanitizer (MSAN), `FinalizeQuantMatrix` triggers a use-of-uninitialized-value warning.

**Root cause:**
1. `uint8_t opt_quants[2][64];` is allocated on the stack uninitialized at the start of `LoopScan()`.
2. If an allocation fails on pass 0 (`p == 0`) inside `StoreRunLevels()`, `if (!ok_) break;` exits the dichotomy loop before `CopyQuantMatrix(quants_[c].quant_, opt_quants[c])` is ever executed.
3. Execution falls through to `SetQuantMatrices(opt_quants)` and `FinalizeQuantMatrix(&quants_[c], q_bias_)`, which evaluates `if (q->quant_[i] < q->min_quant_[i])` on garbage stack memory from `opt_quants`.
4. Additionally, `StoreRunLevels()` and `StoreOptimalHuffmanTables()` on lines 178–183 were being called even when `!ok_`.

**Suggested fix:** Guard the post-search matrix transfer, Huffman optimization, and bitstream generation with `if (ok_)`:

```diff
--- a/src/dichotomy.cc
+++ b/src/dichotomy.cc
@@ -168,24 +168,24 @@ void Encoder::LoopScan() {
     if (search_hook_->Update(result)) break;
   }
-  // transfer back the final matrices
-  SetQuantMatrices(opt_quants);
-  for (int c = 0; c < 2; ++c) FinalizeQuantMatrix(&quants_[c], q_bias_);
-
-  // return informative values to the user
-  search_hook_->q = best_q;
-  search_hook_->value = best_result;
-
-  // optimize Huffman table now, if we haven't already during the search
-  if (!search_hook_->for_size || !last_is_best) {
-    StoreRunLevels(base_coeffs);
-    if (ok_ && optimize_size_) {
-      StoreOptimalHuffmanTables(nb_mbs, base_coeffs);
-    }
-  }
-
-  // finish bitstream
   if (ok_) {
+    // transfer back the final matrices
+    SetQuantMatrices(opt_quants);
+    for (int c = 0; c < 2; ++c) FinalizeQuantMatrix(&quants_[c], q_bias_);
+
+    // return informative values to the user
+    search_hook_->q = best_q;
+    search_hook_->value = best_result;
+
+    // optimize Huffman table now, if we haven't already during the search
+    if (!search_hook_->for_size || !last_is_best) {
+      StoreRunLevels(base_coeffs);
+      if (ok_ && optimize_size_) {
+        StoreOptimalHuffmanTables(nb_mbs, base_coeffs);
+      }
+    }
+
     WriteDQT();
     WriteSOF();
     WriteDHT();
```

Comment (hasinoff) [inline src/dichotomy.cc:20]: Add direct `#include <assert.h>`

File: src/dichotomy.cc [hunk 1]
```diff
@@ -213,7 +215,7 @@ size_t Encoder::HeaderSize() const {
       size += (xmp_.size() / 65458 + 1) * 40;
     }
   }
-  size += 2 * 65 + 2 + 2;         // DQT
+  size += (nb_comps_ == 1 ? 1 : 2) * 65 + 2 + 2;  // DQT
```
Comment (hasinoff) [inline src/dichotomy.cc:223]: Commit 10 fixes the DQT table count from 2 to 1 in `HeaderSize()` for grayscale images (`nb_comps_ == 1`). It would be great to add a test in `tests/unit_test.cc` verifying that `TARGET_SIZE` rate control on `SJPEG_YUV_400` converges accurately to the requested byte target.

File: src/enc.cc
Comment (hasinoff) [inline src/enc.cc]: **Integer Overflow on Valid 16-bit Max Dimensions ($65535 \times 65535$)**

While Commit 6 correctly rejects dimensions $> 65535$, valid large dimensions (e.g. $65535 \times 65535$) cause 32-bit signed integer overflow in intermediate calculations:

1. In `EncoderSharp420`:
   `W * H` evaluates as signed 32-bit `int` ($65535^2 = 4{,}294{,}836{,}225 > \text{INTMAX}$), overflowing to negative values before converting to `size_t`.
   ```cpp
   // Sizing buffer safely:
   const size_t y_size = static_cast<size_t>(W) * H;
   const size_t uv_size = static_cast<size_t>(uv_w) * uv_h;
   yuv_memory_ = Alloc<uint8_t>(y_size + 2 * uv_size);
   if (ok_) {
     y_ = yuv_memory_;
     y_step_ = W;
     u_ = yuv_memory_ + y_size;
     v_ = u_ + uv_size;
     ...
   }
   ```
2. In `SjpegEncode()`:
   ```cpp
   MemorySink sink(static_cast<size_t>(width) * height / 4);
   ```

Comment (hasinoff) [inline src/enc.cc:27]: - Add `#include "bit_writer.h"`: `enc.cc` directly constructs and uses `MemorySink sink(...)`, which is declared in `bit_writer.h` (and not re-exported by `sjpegi.h`).
- Add `#include <algorithm>` for `std::min` / `std::max`.

File: src/jpeg_tools.cc
Comment (hasinoff) [inline src/jpeg_tools.cc:23]: **Heads-up on Grayscale Detection & Direct Includes:**

1. **Grayscale Detection (Commit 9)**: Fixing the `gray_count` ratio calculation in `SjpegRiskiness` works nicely and correctly detects small/flat images as grayscale (`SJPEG_YUV_400`). As a side effect, bitstreams for small flat images drop below 300 bytes (1 component, 1 DQT). Downstream test harnesses and fuzzers corrupting JPEG header bytes should ensure they bound random offset indices to `std::min(jpeg.size(), max_bytes)` rather than indexing against raw RGB buffer dimensions.
2. **Direct Includes**: Add direct `#include <algorithm>` for `std::min` and `#include <stdlib.h>` for `abs`.

File: src/yuv_convert.cc [hunk 2]
```diff
@@ -111,9 +112,9 @@ static uint32_t kLinearToGammaTab[GAMMA_TABLE_SIZE + 2];
 static uint32_t kGammaToLinearTab[MAX_Y_T + 1];   // size scales with Y_FIX
 
 static void InitGammaTablesF() {
-  static bool done = false;
+  static std::once_flag once;
   assert(2 * GAMMA_TO_LINEAR_BITS < 32);  // we use uint32_t intermediate values
-  if (!done) {
+  std::call_once(once, []() {
```
Comment (hasinoff) [inline src/yuv_convert.cc:117]: **Test Coverage Suggestion (Commit 11)**

Using `static std::once_flag` and `std::call_once` ensures thread safety for the lazy lookup tables. It would be helpful to add a test in `tests/unit_test.cc` that spawns multiple concurrent `std::thread` workers simultaneously calling `SjpegCompress` with `SJPEG_YUV_SHARP` to verify race-free execution under ThreadSanitizer.

File: tests/unit_test.cc
Comment (hasinoff) [inline tests/unit_test.cc:1]: **Empty Vector UB, Direct Includes & Test Additions:**
1. **Direct Includes**:
   - Add `#include <algorithm>` for `std::min` / `std::max`.
2. **Empty Vector UB**: In `EncodeRGB()` and `InvalidArguments()`, prefer `vec.data()` over `&vec[0]` to avoid undefined behavior on empty vectors in C++11.
3. **Negative Strides & Odd Dimensions**: Consider adding test cases with negative strides (`stride < 0` for vertically flipped input buffers) in `TEST(EncodeYUV420Strides)` and `TEST(EncodeNV)`, as well as odd dimensions (e.g. $17 \times 13$) to verify sub-sampling boundary clipping.

**Review by hasinoff — COMMENTED** [review]
Thanks for putting together this extensive cleanup!

We ran the entire 14-commit series through extensive testing, including our full integration test suites, high-iteration fuzzing (**5,000,000 iterations** under ASAN `libFuzzer`), and multi-sanitizer stress testing (**ASAN**, **MSAN**, **UBSAN**, **TSAN**).

All core functionality compiles cleanly and tests pass with zero regressions. In our inline comments below, we have noted:
1. **A MemorySanitizer (MSAN) uninitialized stack read defect** in `src/dichotomy.cc` during allocation failure handling (`TEST(AllocationFailure)`), along with a patch.
2. **Integer overflow edge cases** in `src/enc.cc` when evaluating valid max dimensions ($65535 \times 65535$).
3. **Missing direct includes** (`bit_writer.h` for `MemorySink`, `<assert.h>`, `<algorithm>`).
4. **Suggested test additions** in `tests/unit_test.cc` to cover single-component grayscale rate control, multithreaded `std::call_once` lazy initialization, and negative strides.

File: src/dichotomy.cc
Comment (skal-aa) [reply src/dichotomy.cc]: Fixed in 608a89c, using your patch -- except that the bitstream writes keep their own inner `if (ok_)`: the `StoreRunLevels()` call inside the guarded block can still clear `ok_` through `CheckBuffers()`.

Comment (skal-aa) [reply src/dichotomy.cc]: Added.

File: src/dichotomy.cc (hunk 1 above)
Comment (skal-aa) [reply src/dichotomy.cc]: Added `TEST(TargetSize)`: `SJPEG_YUV_400` converges within 1.4%, same as `SJPEG_YUV_420`, against 4.8-8.4% before the fix.

File: src/enc.cc
Comment (skal-aa) [reply src/enc.cc]: Fixed: the plane sizes are computed as `size_t`, and all six `width * height / 4` hints are cast now (only two were). The sharp converter is still int-indexed internally, so 65535x65535 needs a separate pass.

Comment (skal-aa) [reply src/enc.cc]: `bit_writer.h` added, along with `<string.h>` and `<string>`. `<algorithm>` is not needed: `enc.cc` has no `std::min` or `std::max`, and the `std::abs` it does use comes from `<cstdlib>`.

File: src/jpeg_tools.cc
Comment (skal-aa) [reply src/jpeg_tools.cc]: Noted on the fuzzer bounds, thanks. Added `<utility>` for the `std::swap` at line 193; `<algorithm>` and `<stdlib.h>` are not needed, as this file has no `min`, `max` or `abs`.

File: src/yuv_convert.cc (hunk 2 above)
Comment (skal-aa) [reply src/yuv_convert.cc]: Added `TEST(Threads)`: eight concurrent `SJPEG_YUV_SHARP` encodes, which TSAN flags as a race if the `call_once` is reverted.

File: tests/unit_test.cc
Comment (skal-aa) [reply tests/unit_test.cc]: Switched to `.data()`, and added `TEST(NegativeStrides)` (RGB / YUV420 / NV12 against the flipped picture) plus 17x13 to the U/V stride tests. `<algorithm>` is not needed here either.

File: src/dichotomy.cc
Comment (skal65535) [reply src/dichotomy.cc]: Fixed in 608a89c, using your patch -- except that the bitstream writes keep their own inner `if (ok_)`: the `StoreRunLevels()` call inside the guarded block can still clear `ok_` through `CheckBuffers()`.

Comment (skal65535) [reply src/dichotomy.cc:20]: Added.

File: src/dichotomy.cc (hunk 1 above)
Comment (skal65535) [reply src/dichotomy.cc:223]: Added `TEST(TargetSize)`: `SJPEG_YUV_400` converges within 1.4%, same as `SJPEG_YUV_420`, against 4.8-8.4% before the fix.

File: src/enc.cc
Comment (skal65535) [reply src/enc.cc]: Fixed: the plane sizes are computed as `size_t`, and all six `width * height / 4` hints are cast now (only two were). The sharp converter is still int-indexed internally, so 65535x65535 needs a separate pass.

Comment (skal65535) [reply src/enc.cc:27]: `bit_writer.h` added, along with `<string.h>` and `<string>`. `<algorithm>` is not needed: `enc.cc` has no `std::min` or `std::max`, and the `std::abs` it does use comes from `<cstdlib>`.

File: src/jpeg_tools.cc
Comment (skal65535) [reply src/jpeg_tools.cc:23]: Noted on the fuzzer bounds, thanks. Added `<utility>` for the `std::swap` at line 193; `<algorithm>` and `<stdlib.h>` are not needed, as this file has no `min`, `max` or `abs`.

File: src/yuv_convert.cc (hunk 2 above)
Comment (skal65535) [reply src/yuv_convert.cc:117]: Added `TEST(Threads)`: eight concurrent `SJPEG_YUV_SHARP` encodes, which TSAN flags as a race if the `call_once` is reverted.

File: tests/unit_test.cc
Comment (skal65535) [reply tests/unit_test.cc:1]: Switched to `.data()`, and added `TEST(NegativeStrides)` (RGB / YUV420 / NV12 against the flipped picture) plus 17x13 to the U/V stride tests. `<algorithm>` is not needed here either.

**Comment by skal65535** [comment]
@hasinoff , i fixed all your comments , thanks for the testing! PTAL.

**Comment by hasinoff** [comment]
LGTM

File: src/jpeg_tools.cc [hunk 3]
```diff
@@ -31,27 +32,28 @@ namespace {
...
-  const uint8_t* const end = src + size - 8;   // 8 bytes of safety, for marker
-  src += 2;   // skip M_SOI
-  for (; src < end && *src != 0xff; ++src) { /* search first 0xff marker */ }
-  while (src < end) {
-    const uint32_t marker = static_cast<uint32_t>((src[0] << 8) | src[1]);
-    if (marker == M_SOF0 || marker == M_SOF1) return src;
-    const size_t s = 2 + ((src[2] << 8) | src[3]);
-    src += s;
+// Positions are used instead of pointers, to never build one past the end.
+const uint8_t* GetSOFData(const uint8_t* src, size_t size) {
```
Comment (jzern) [inline src/jpeg_tools.cc:36]: Work like this has been transitioning to `std::span`.

File: CMakeLists.txt [hunk 4]
```diff
@@ -27,6 +27,7 @@ set(CMAKE_CXX_STANDARD 11)
 # Options for coder / decoder executables.
 option(SJPEG_ENABLE_SIMD "Enable any SIMD optimization." ON)
 option(SJPEG_BUILD_EXAMPLES "Build the sjpeg / vjpeg command line tools." ON)
+option(SJPEG_BUILD_TESTS "Build the unit tests." ON)
```
Comment (jzern) [inline CMakeLists.txt:30]: Note `include(CTest)` provides [`BUILD_TESTING`](https://cmake.org/cmake/help/latest/variable/BUILD_TESTING.html).

Comment (skal-aa) [reply CMakeLists.txt:30]: ah, right. TODO(later)

File: src/jpeg_tools.cc (hunk 3 above)
Comment (skal-aa) [reply src/jpeg_tools.cc:36]: that's C++20 feature, and we're still using C++11 here (probably i bad idea overall)
